src/x86_64/schedule.c

```diff
@@ -0,0 +1,182 @@
+#include <kernel.h>
+#include <tfs.h> // needed for unix.h
+#include <unix.h> // some deps
+#include <apic.h>
+
+
+/* Try to keep these within the confines of the runloop lock so we
+   don't create too much of a mess. */
+//#define SCHED_DEBUG
+#ifdef SCHED_DEBUG
+#define sched_debug(x, ...) do {log_printf("SCHED", "[%2d] " x, ci->id, ##__VA_ARGS__);} while(0)
+#else
+#define sched_debug(x, ...) (void)ci
+#endif
+
+// currently defined in x86_64.h
+static char *state_strings_backing[] = {
+    "not present",
+    "idle",
+    "kernel",
+    "interrupt",
+    "user",         
+};
+
+char **state_strings = state_strings_backing;
+static int wakeup_vector;
+int shutdown_vector;
+
+queue runqueue;                 /* kernel space from ?*/
+queue bhqueue;                  /* kernel from interrupt */
+queue thread_queue;             /* kernel to user */
+timerheap runloop_timers;
+u64 idle_cpu_mask;              /* xxx - limited to 64 aps. consider merging with bitmask */
+timestamp last_timer_update;
+
+static timestamp runloop_timer_min;
+static timestamp runloop_timer_max;
+
+static u64 kernel_lock;
+
+void kern_lock()
+{
+    cpuinfo ci = current_cpu();
+    spin_lock(&kernel_lock);
+    ci->have_kernel_lock = true;
+}
+
+boolean kern_try_lock()
+{
+    cpuinfo ci = current_cpu();
+    if (ci->have_kernel_lock)
+        return true;
+    if (!spin_try(&kernel_lock))
+        return false;
+    ci->have_kernel_lock = true;
+    return true;
+}
+
+void kern_unlock()
+{
+    cpuinfo ci = current_cpu();
+    assert(ci->have_kernel_lock);
+    ci->have_kernel_lock = false;
+    spin_unlock(&kernel_lock);
+}
+
+static void run_thunk(thunk t, int cpustate)
+{
+    cpuinfo ci = current_cpu();
+    sched_debug(" run: %F state: %s\n", t, state_strings[cpustate]);
+    // as we are walking by, if there is work to be done and an idle cpu,
+    // get it to wake up and examine the queue
+    if (idle_cpu_mask &&
+        ((queue_length(bhqueue) > 0) ||
+         (queue_length(runqueue) > 0) ||
+         (queue_length(thread_queue) > 0))) {
+        // unfortunately, if idle cpu mask is zero (which can happen since this
+        // is racy), the result is the previous value ... so asm here
+        u64 mask_copy = idle_cpu_mask;
+        u64 cpu = msb(mask_copy);        
+        // this really shouldn't ever be current_cpu() ? 
+        if (cpu != INVALID_PHYSICAL && cpu != current_cpu()->id) {
+            sched_debug("sending wakeup ipi to %d %x\n", cpu, wakeup_vector);
+            atomic_clear_bit(&idle_cpu_mask, cpu);
+            apic_ipi(cpu, 0, wakeup_vector);
+        }
+    }
+
+    ci->state = cpustate;
+    apply(t);
+    // do we want to enforce this? i kinda just want to collapse
+    // the stack and ensure that the thunk actually wanted to come back here
```

Comment:
I think actually most kernel thunks do just return - unless it's resuming a syscall (though that might just return too after scheduling the thread - but it might become a frame return in the fast path, as discussed earlier).

The thread thunks, on the other hand, shouldn't return...